### PR TITLE
Initial Implementation of Databricks HealthCheck

### PR DESCRIPTION
**Pull request by ftoft**
## Description

We have implemented a health check for presister streaming job and publisher streming job.

## References

* #406

### CODE REVIEW

**Review by BjarkeMeier — COMMENTED** [review]
A few preliminary comments:
We'll most likely want to reuse this in wholesale. So perhaps we should already now consider providing (and using) this as a NuGet package.
Our code conventions suggest only one type per file.

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs [hunk 1]
```diff
@@ -110,6 +110,9 @@ protected override void OnConfigureEnvironment()
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.TimeSeriesRaw, TimeSeriesRawFolderName);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.B2CTenantId, AuthorizationConfiguration.B2cTenantId);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.BackendServiceAppId, AuthorizationConfiguration.BackendAppId);
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksApiToken, "INSERT TOKEN HERE");
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksApiUri, "https://adb-5870161604877074.14.azuredatabricks.net");
```
Comment (lasrinnil) [inline source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs:114]: "https://adb-5870161604877074.14.azuredatabricks.net" could you move this URI to a constant and give it a name for the reader to have an easier time why its valid?

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs [hunk 2]
```diff
@@ -110,6 +110,9 @@ protected override void OnConfigureEnvironment()
...
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksHealthCheckEnabled, "False");
```
Comment (lasrinnil) [inline source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs:115]: why is this required?

File: source/TimeSeries/TestCore/TestCore.csproj [hunk 3]
```diff
@@ -34,6 +34,8 @@ limitations under the License.
     <EmbeddedResource Include="Assets\Invalid_Hourly_CIM_TimeSeries_missing_mRID.xml" />
     <EmbeddedResource Include="Assets\Valid_Hourly_CIM_MultipleTimeSeries.xml" />
     <EmbeddedResource Include="Assets\Valid_Hourly_CIM_TimeSeries.xml" />
+    <None Remove="Assets\DatabricksClustersListResponse.json" />
```
Comment (lasrinnil) [inline source/TimeSeries/TestCore/TestCore.csproj:37]: remove

File: source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    private readonly HttpClient _httpClient;
+
+    public DatabricksClient(HttpClient httpClient)
+    {
+        _httpClient = httpClient;
+    }
+
+    public async Task<Root?> GetClustersAsync()
+    {
+        using var response = await _httpClient.GetAsync("api/2.0/clusters/list");
```
Comment (lasrinnil) [inline source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs:34]: I know the client used in the wholesale domain is version 2.1. 

Could you perhaps test if that version exposes what is required to perform the healthcheck

File: source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        response.EnsureSuccessStatusCode();
+        await using var stream = await response.Content.ReadAsStreamAsync();
+        return JsonSerializer.Deserialize<Root>(stream);
+    }
+}
+
+public class Root
```
Comment (lasrinnil) [inline source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs:41]: Please move classes to separate files

File: source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    public string? ClusterId { get; set; }
+
+    [JsonPropertyName("state")]
+    public string? State { get; set; }
+
+    [JsonPropertyName("default_tags")]
+    public DefaultTags? DefaultTags { get; set; }
+}
+
+public class DefaultTags
```
Comment (lasrinnil) [inline source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs:59]: Is `DefaultTags` valid if all the strings are null?

File: source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+{
+    public string? ClusterName { get; set; }
+
+    public string? JobId { get; set; }
+
+    public string? RunName { get; set; }
```
Comment (lasrinnil) [inline source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs:65]: What is `RunName`? Perhaps a description would help the reader to understand (and me)

File: source/TimeSeries/TimeSeriesBundleIngestor/DatabricksHealthCheck.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            response = await _client.GetClustersAsync();
+        }
+        catch (HttpRequestException e)
+        {
+            return HealthCheckResult.Unhealthy("Was not able to get a successfully response from Databricks API StatusCode: " + e.StatusCode + " - " + e.Message);
+        }
+
+        if (response?.Clusters == null)
+            return HealthCheckResult.Unhealthy("Was not able to get a successfully response from Databricks API");
```
Comment (lasrinnil) [inline source/TimeSeries/TimeSeriesBundleIngestor/DatabricksHealthCheck.cs:52]: if clusters are not present. Consider returning a message informing of that, instead of a more generic message.

File: source/TimeSeries/TimeSeriesBundleIngestor/IDatabricksClient.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+
+namespace Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor;
+
+public interface IDatabricksClient
+{
+    Task<Root?> GetClustersAsync();
```
Comment (lasrinnil) [inline source/TimeSeries/TimeSeriesBundleIngestor/IDatabricksClient.cs:21]:
```suggestion
    Task<Root?> GetRootAsync();
```

File: source/TimeSeries/UnitTests/DatabricksClientTests.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    public DatabricksClientTests()
+    {
+        _testDocuments = new TestDocuments();
+    }
+
+    [Fact]
+    public async Task SuccessfulGetClustersRequest_ParseSuccessful()
+    {
+        var mockHttp = new MockHttpMessageHandler();
```
Comment (lasrinnil) [inline source/TimeSeries/UnitTests/DatabricksClientTests.cs:41]: Could you consider simply using the Moq framework to mock the response from the Httpclient?

I've found this article describing an approach:
https://gingter.org/2018/07/26/how-to-mock-httpclient-in-your-net-c-unit-tests/

File: build/infrastructure/main/func-time-series-bundle-ingestor.tf [hunk 11]
```diff
@@ -43,6 +43,11 @@ module "time_series_bundle_ingestor" {
     EVENT_HUB_NAME                                      = module.evh_received_timeseries.name
     "TimeSeriesRawFolder:FolderName"                    = local.DATA_LAKE_TIME_SERIES_RAW_FOLDER_NAME
     DATA_LAKE_CONTAINER_NAME                            = local.DATA_LAKE_CONTAINER_NAME
+    DATABRICKS_API_TOKEN                                = "@Microsoft.KeyVault(VaultName=${var.shared_resources_keyvault_name};SecretName=dbw-shared-workspace-token)"
+    DATABRICKS_API_URI                                  = "@Microsoft.KeyVault(VaultName=${var.shared_resources_keyvault_name};SecretName=dbw-shared-workspace-url)"
+    DATABRICKS_PRESISTER_STREAMING_JOB_NAME             = databricks_job.persister_streaming_job.name
```
Comment (lasrinnil) [inline build/infrastructure/main/func-time-series-bundle-ingestor.tf:48]:
```suggestion
    DATABRICKS_PERSISTER_STREAMING_JOB_NAME             = databricks_job.persister_streaming_job.name
```

File: source/TimeSeries/UnitTests/DatabricksClientTests.cs (hunk 10 above)
Comment (johevemi) [reply source/TimeSeries/UnitTests/DatabricksClientTests.cs]: The framework has been replaced with the suggested method using moq

File: source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        });
+
+        var sut = new DatabricksHealthCheck(new[] { "persister_streaming_job", "publisher_streaming_job" }, databricksClientMock.Object, loggerMock.Object);
+
+        var result = await sut.CheckHealthAsync(new HealthCheckContext());
+        Assert.Equal(healthStatus, result.Status);
+    }
+
+    [Fact]
+    public async Task GivenDatabricksHasBadToken_Return_UnHealthy()
```
Comment (lasrinnil) [inline source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs:74]: Where is the bad token given / used?

**Comment by lasrinnil** [comment]
I have completed my review, please reach out to me for any further clarification.

**Review by johevemi — APPROVED** [review]

File: source/TimeSeries/TestCore/TestCore.csproj (hunk 3 above)
Comment (johevemi) [reply source/TimeSeries/TestCore/TestCore.csproj]: has been removed

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs (hunk 1 above)
Comment (johevemi) [reply source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs]: moved to private const

File: source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs (hunk 5 above)
Comment (johevemi) [reply source/TimeSeries/TimeSeriesBundleIngestor/DatabricksClient.cs]: done

**Comment by johevemi** [comment]
I have verified that we get a current health status and that we get a Teams notification through application insight. note: It takes about 15 min from receiving the unhealthy message in application insight to receive the message in Teams

**Comment by BjarkeMeier** [comment]
> I have verified that we get a current health status and that we get a Teams notification through application insight. note: It takes about 15 min from receiving the unhealthy message in application insight to receive the message in Teams

Why that long? Shouldn't it be within around 5 minutes?

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs [hunk 13]
```diff
@@ -110,6 +112,9 @@ protected override void OnConfigureEnvironment()
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.TimeSeriesRaw, TimeSeriesRawFolderName);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.B2CTenantId, AuthorizationConfiguration.B2cTenantId);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.BackendServiceAppId, AuthorizationConfiguration.BackendAppId);
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksApiToken, "INSERT TOKEN HERE");
```
Comment (BjarkeMeier) [inline source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs:115]: When and why should who insert a token? And where should that token be obtained?

File: source/TimeSeries/TestCore/Assets/TestDocuments.cs [hunk 14]
```diff
@@ -34,6 +34,8 @@ public class TestDocuments
 
         public string TimeSeriesBundleJsonAsString => GetDocumentAsString("TimeSeriesBundleJson.Time_Series_Bundle.json");
 
+        public string DatabricksClustersListResponse => GetDocumentAsString("DatabricksClustersListResponse.json");
+
         public string ValidMultipleTimeSeriesAsStringWithGuid(string baseFileName) => GetDocumentAsStringReplaceIdWithGuid("Valid_Hourly_CIM_MultipleTimeSeries.xml", baseFileName, "C1876453");
```
Comment (BjarkeMeier) [inline source/TimeSeries/TestCore/Assets/TestDocuments.cs:39]: Looks like we are building an ever growing list of different types of test files. Could we make subfolders for e.g.
- time series CIM XML documents
- Databricks API request/responses
- Perhaps other?

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Cluster.cs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+
+using System.Text.Json.Serialization;
+
+namespace Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor.Monitor.Databricks;
+
+public class Cluster
+{
+    [JsonPropertyName("cluster_id")]
+    public string? ClusterId { get; set; }
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Cluster.cs:22]: Can the ID be null?

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheck.cs [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            response = await _healthCheckClient.GetClustersAsync();
+        }
+        catch (HttpRequestException e)
+        {
+            return HealthCheckResult.Unhealthy("Was not able to get a successfully response from Databricks API StatusCode: " + e.StatusCode + " - " + e.Message);
+        }
+
+        if (response?.Clusters == null)
+            return HealthCheckResult.Unhealthy("Was not able to get a get clusters from Databricks API");
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheck.cs:52]:
```suggestion
            return HealthCheckResult.Unhealthy("Failed to get clusters from Databricks API");
```

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheck.cs [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                runningJobs.Add(jobName);
+            }
+            else
+            {
+                notRunningJobs.Add(jobName);
+            }
+        }
+
+        if (!notRunningJobs.Any())
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheck.cs:71]: Perhaps invert the if-statement by switching the branches and remove the not operator for readability?

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheck.cs [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            return HealthCheckResult.Healthy($@"Running jobs {string.Join(", ", runningJobs)}");
+        }
+        else
+        {
+            return HealthCheckResult.Unhealthy($@"Not running jobs {string.Join(", ", notRunningJobs)} and running jobs {string.Join(", ", runningJobs)}");
+        }
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheck.cs:78]:
```suggestion
        return HealthCheckResult.Unhealthy($@"Not running jobs {string.Join(", ", notRunningJobs)} and running jobs {string.Join(", ", runningJobs)}");
```

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheckClient.cs [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+public class DatabricksHealthCheckClient : IDatabricksHealthCheckClient
+{
+    private readonly HttpClient _httpClient;
+
+    public DatabricksHealthCheckClient(HttpClient httpClient)
+    {
+        _httpClient = httpClient;
+    }
+
+    public async Task<Root?> GetClustersAsync()
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheckClient.cs:30]: Why do we return a "Root" instead of a collection of clusters?

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DefaultTags.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor.Monitor.Databricks;
+
+public class DefaultTags
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DefaultTags.cs:17]: Does these tags apply to multiple different entities? Or should they perhaps be named `JobDefaultTags` or...?

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Root.cs [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+
+namespace Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor.Monitor.Databricks;
+
+public class Root
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Root.cs:20]: What is a root?

File: source/TimeSeries/TimeSeriesBundleIngestor/Program.cs [hunk 22]
```diff
@@ -96,6 +98,14 @@ private static void ConfigureServices(IServiceCollection serviceCollection)
                 return storage;
             });
             serviceCollection.AddScoped<RequestResponseLoggingMiddleware>();
+            serviceCollection.AddSingleton<IDatabricksHealthCheckClient>(x =>
+            {
+                var httpClient = new HttpClient();
+                httpClient.BaseAddress = new Uri("https://" + EnvironmentHelper.GetEnv(EnvironmentSettingNames.DatabricksApiUri));
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Program.cs:104]: Why isn't "https://" part of the env url variable?

File: source/TimeSeries/TimeSeriesBundleIngestor/Program.cs [hunk 23]
```diff
@@ -104,6 +114,11 @@ private static void ConfigureServices(IServiceCollection serviceCollection)
                 .AddAzureEventHub(name: "EventhubConnectionExists", eventHubConnectionFactory: options => new EventHubConnection(
                     EnvironmentHelper.GetEnv("EVENT_HUB_CONNECTION_STRING"),
                     EnvironmentHelper.GetEnv("EVENT_HUB_NAME")));
+
+            if (EnvironmentHelper.GetEnv(EnvironmentSettingNames.DatabricksHealthCheckEnabled).ToUpper() == "TRUE")
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Program.cs:118]: Should we catch misspelled values like e.g. "fallse"? E.g. throwing an exception?

File: source/TimeSeries/TimeSeriesBundleIngestor/Program.cs (hunk 22 above)
Comment (ftoft) [reply source/TimeSeries/TimeSeriesBundleIngestor/Program.cs:104]: Its not part of the value we get from terraform shared keyvault.

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Root.cs (hunk 21 above)
Comment (ftoft) [reply source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Root.cs:20]: Its the root of the data structure we get from the api.

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DefaultTags.cs (hunk 20 above)
Comment (ftoft) [reply source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DefaultTags.cs:17]: Its because it's the name of the object in the json structure.

File: source/TimeSeries/TimeSeriesBundleIngestor/Program.cs (hunk 23 above)
Comment (lasrinnil) [reply source/TimeSeries/TimeSeriesBundleIngestor/Program.cs]: I have created a private method for determining if the eviormentsetting is either TRUE or FALSE or throws an exception

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Root.cs (hunk 21 above)
Comment (lasrinnil) [reply source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Root.cs:20]: Perhaps we could call it something more specific like `DatabricksClusterRespons` ?

File: source/TimeSeries/TimeSeriesBundleIngestor/Program.cs (hunk 22 above)
Comment (BjarkeMeier) [reply source/TimeSeries/TimeSeriesBundleIngestor/Program.cs:104]: I think we should fix the problem at the root - which means we should fix the value in the shared keyvault instead of having to debug and workaround in all consuming apps.

File: source/TimeSeries/TestCore/Assets/TestDocuments.cs (hunk 14 above)
Comment (johevemi) [reply source/TimeSeries/TestCore/Assets/TestDocuments.cs:39]: wont move at this time

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs [hunk 24]
```diff
@@ -37,6 +37,8 @@ public class TimeSeriesFunctionAppFixture : FunctionAppFixture
 
         private const string MarketOpLogs = "marketoplogs";
 
+        private const string EuWestDatabricksWorkspaceUrl = "https://adb-5870161604877074.14.azuredatabricks.net";
```
Comment (BjarkeMeier) [inline source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs:40]: This should be obtained from keyvault.

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs (hunk 13 above)
Comment (BjarkeMeier) [inline source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs:115]:
```suggestion
            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksApiToken, "THIS IS A DUMMY TOKEN");
```

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DefaultTags.cs (hunk 20 above)
Comment (lasrinnil) [reply source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DefaultTags.cs:17]: It is renamed to DefaultTags to follow the API

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs [hunk 25]
```diff
@@ -110,6 +112,9 @@ protected override void OnConfigureEnvironment()
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.TimeSeriesRaw, TimeSeriesRawFolderName);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.B2CTenantId, AuthorizationConfiguration.B2cTenantId);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.BackendServiceAppId, AuthorizationConfiguration.BackendAppId);
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksApiToken, "INSERT TOKEN HERE");
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksApiUri, EuWestDatabricksWorkspaceUrl);
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DatabricksHealthCheckEnabled, "False");
```
Comment (BjarkeMeier) [inline source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs:117]: Just a though: We have an `EnvironmentVariableHelper`. How about using that everywhere and make overloads for get and set that manages booleans?

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheckClient.cs (hunk 19 above)
Comment (lasrinnil) [reply source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/DatabricksHealthCheckClient.cs]: We now return a list of clusters wrapped in a `DatabricksClusterListAsync`

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Cluster.cs (hunk 15 above)
Comment (lasrinnil) [reply source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Cluster.cs]: Removed the property it had no longer any relevance for the healthcheck

File: source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Cluster.cs [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor.Monitor.Databricks;
+
+public class Cluster
+{
+    [JsonPropertyName("state")]
+    public string? State { get; set; }
+
+    [JsonPropertyName("default_tags")]
+    public DefaultTags? DefaultTags { get; set; }
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/Monitor/Databricks/Cluster.cs:25]:
```suggestion
    public JobDefaultTags? DefaultTags { get; set; }
```

File: source/TimeSeries/TimeSeriesBundleIngestor/IDatabricksHealthCheckClient.cs [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+// limitations under the License.
+
+using System.Threading.Tasks;
+using Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor.Monitor.Databricks;
+
+namespace Energinet.DataHub.TimeSeries.TimeSeriesBundleIngestor;
+
+public interface IDatabricksHealthCheckClient
+{
+    Task<DatabricksClusterListResponse?> GetClusterListAsync();
```
Comment (BjarkeMeier) [inline source/TimeSeries/TimeSeriesBundleIngestor/IDatabricksHealthCheckClient.cs:22]:
```suggestion
    Task<List<Cluster>> GetClustersAsync();
```

File: source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+{
+    [Theory]
+    [InlineData("RUNNING", "RUNNING", HealthStatus.Healthy)]
+    [InlineData("TERMINATED", "TERMINATED", HealthStatus.Unhealthy)]
+    [InlineData("RUNNING", "TERMINATED", HealthStatus.Unhealthy)]
+    public async Task GivenTwoDatabricksTaskRuns_Return_CorrectHealthStatus(string state1, string state2, HealthStatus healthStatus)
+    {
+        var loggerMock = new Mock<ILogger<DatabricksHealthCheck>>();
+        var databricksClientMock = new Mock<IDatabricksHealthCheckClient>();
+        databricksClientMock.Setup(x => x.GetClusterListAsync()).ReturnsAsync(new DatabricksClusterListResponse
```
Comment (BjarkeMeier) [inline source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs:39]: The test will be much more readable if you move the response creation into a separate method.

File: source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+    [Fact]
+    public async Task GivenDatabricksHasBadToken_Return_UnHealthy()
+    {
+        var loggerMock = new Mock<ILogger<DatabricksHealthCheck>>();
+        var databricksClientMock = new Mock<IDatabricksHealthCheckClient>();
+        databricksClientMock.Setup(x => x.GetClusterListAsync()).Throws(new HttpRequestException());
+
+        var sut = new DatabricksHealthCheck(new string[] { }, databricksClientMock.Object);
+
+        var result = await sut.CheckHealthAsync(new HealthCheckContext());
```
Comment (BjarkeMeier) [inline source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs:81]: nit: I think our usual naming convention is "actual" instead of "result".

File: source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs (hunk 28 above)
Comment (ftoft) [reply source/TimeSeries/UnitTests/DatabricksHealthCheckTests.cs:39]: I not agree with that. i am setting up the test witch is a part of the main purpose of the test. I like to have all data releated to one test in the single method because it is more easy to read, no need to navigate around to understand the test.

File: source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs (hunk 25 above)
Comment (ftoft) [reply source/TimeSeries/IntegrationTests/Fixtures/TimeSeriesFunctionAppFixture.cs]: Switched to use bool.FalseString insted.
